File: backend/src/models/Enquiry.ts
import mongoose, { Types } from 'mongoose'

/**
 * A student's class enquiry against a specific mentor — the new front door
 * (replaces self-service booking). The admin contacts the parties offline and
 * then creates the actual Booking from this record (status → converted).
 */
export interface IEnquiry {
  studentName: string
  email: string
  phone: string
  mentorId: Types.ObjectId
  mentorName?: string // denormalized for the admin list + emails
  selectedSyllabus?: string
  classId?: Types.ObjectId | null
  className?: string
  enquiryType: 'demo' | 'subject-wise'
  subjects: { subjectId?: Types.ObjectId | null; name: string; price: number }[]
  estimatedAmount: number // 0 for a demo
  message?: string
  status: 'new' | 'contacted' | 'converted' | 'closed'
  bookingId?: Types.ObjectId | null // set when the admin converts it
  createdAt?: Date
  updatedAt?: Date
}

const EnquirySchema = new mongoose.Schema<IEnquiry>(
  {
    studentName: { type: String, required: true, trim: true },
    email: { type: String, required: true, lowercase: true, trim: true },
    phone: { type: String, required: true, trim: true },
    mentorId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'AuthModal',
      required: true,
    },
    mentorName: { type: String, default: '' },
    selectedSyllabus: { type: String, default: '' },
    classId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'ClassesModel',
      default: null,
    },
    className: { type: String, default: '' },
    enquiryType: {
      type: String,
      enum: ['demo', 'subject-wise'],
      required: true,
    },
    subjects: [
      {
        subjectId: {
          type: mongoose.Schema.Types.ObjectId,
          ref: 'SubjectModel',
          default: null,
        },
        name: { type: String, default: '' },
        price: { type: Number, default: 0 },
      },
    ],
    estimatedAmount: { type: Number, default: 0 },
    message: { type: String, default: '' },
    status: {
      type: String,
      enum: ['new', 'contacted', 'converted', 'closed'],
      default: 'new',
    },
    bookingId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Booking',
      default: null,
    },
  },
  { timestamps: true }
)

// Admin list: recent-first, filterable by status.
EnquirySchema.index({ status: 1, createdAt: -1 })
EnquirySchema.index({ mentorId: 1, createdAt: -1 })

export default mongoose.model('Enquiry', EnquirySchema)

File: scholarhub_mobile/lib/data/models/enquiry.dart
// Class-enquiry client model (the lead front-door that replaces booking).

num _num(dynamic v) {
  if (v is num) return v;
  if (v is String) return num.tryParse(v) ?? 0;
  return 0;
}

DateTime? _date(dynamic v) => v is String ? DateTime.tryParse(v) : null;
String _str(dynamic v) => (v ?? '').toString();

enum EnquiryType { demo, subjectWise }

extension EnquiryTypeX on EnquiryType {
  String get apiValue =>
      this == EnquiryType.demo ? 'demo' : 'subject-wise';
  String get label => this == EnquiryType.demo ? 'Demo class' : 'Subject-wise';
}

class EnquirySubject {
  final String? subjectId;
  final String name;
  final num price;

  const EnquirySubject({
    required this.subjectId,
    required this.name,
    required this.price,
  });

  Map<String, dynamic> toJson() => {
        'subjectId': ?subjectId,
        'name': name,
        'price': price,
      };

  factory EnquirySubject.fromJson(Map<String, dynamic> json) => EnquirySubject(
        subjectId: json['subjectId'] == null ? null : _str(json['subjectId']),
        name: _str(json['name']),
        price: _num(json['price']),
      );
}

/// An enquiry as returned by the admin list.
class EnquiryRecord {
  final String id;
  final String studentName;
  final String email;
  final String phone;
  final String mentorId;
  final String mentorName;
  final String selectedSyllabus;
  final String? classId;
  final String className;
  final EnquiryType enquiryType;
  final List<EnquirySubject> subjects;
  final num estimatedAmount;
  final String message;
  final String status; // new | contacted | converted | closed
  final DateTime? createdAt;

  const EnquiryRecord({
    required this.id,
    required this.studentName,
    required this.email,
    required this.phone,
    required this.mentorId,
    required this.mentorName,
    required this.selectedSyllabus,
    required this.classId,
    required this.className,
    required this.enquiryType,
    required this.subjects,
    required this.estimatedAmount,
    required this.message,
    required this.status,
    required this.createdAt,
  });

  bool get isDemo => enquiryType == EnquiryType.demo;
  bool get isConverted => status == 'converted';

  factory EnquiryRecord.fromJson(Map<String, dynamic> json) {
    final rawSubjects = json['subjects'];
    final subjects = <EnquirySubject>[];
    if (rawSubjects is List) {
      for (final s in rawSubjects) {
        if (s is Map) {
          subjects.add(EnquirySubject.fromJson(s.cast<String, dynamic>()));
        }
      }
    }
    return EnquiryRecord(
      id: _str(json['_id']),
      studentName: _str(json['studentName']),
      email: _str(json['email']),
      phone: _str(json['phone']),
      mentorId: _str(json['mentorId']),
      mentorName: _str(json['mentorName']),
      selectedSyllabus: _str(json['selectedSyllabus']),
      classId: json['classId'] == null ? null : _str(json['classId']),
      className: _str(json['className']),
      enquiryType: _str(json['enquiryType']) == 'demo'
          ? EnquiryType.demo
          : EnquiryType.subjectWise,
      subjects: subjects,
      estimatedAmount: _num(json['estimatedAmount']),
      message: _str(json['message']),
      status: _str(json['status']).isEmpty ? 'new' : _str(json['status']),
      createdAt: _date(json['createdAt']),
    );
  }
}
